File: lintcode/127_topological-sorting/topological-sorting.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/topological-sorting
@Language: Python
@Datetime: 16-05-05 03:15
'''

# Definition for a Directed graph node
# class DirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []


class Solution:
    """
    @param graph: A list of Directed graph node
    @return: A list of graph nodes in topological order.
    """
    def topSort(self, graph):
        def util(node, graph, visited, res):
            visited[node.label] = True
            for n in node.neighbors:
                if not visited[n.label]:
                    util(n, graph, visited, res)
            res.append(node)
        v_size = len(graph)
        visited = [False] * v_size
        res = []
        for node in graph:
            if not visited[node.label]:
                util(node, graph, visited, res)
        return res[::-1]

File: lintcode/127_topological-sorting/topological-sorting.cpp
/*
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/topological-sorting
@Language: C++
@Datetime: 16-05-05 03:09
*/

/**
 * Definition for Directed graph.
 * struct DirectedGraphNode {
 *     int label;
 *     vector<DirectedGraphNode *> neighbors;
 *     DirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    /**
     * @param graph: A list of Directed graph node
     * @return: Any topological order for the given graph.
     */
    vector<DirectedGraphNode*> topSort(vector<DirectedGraphNode*> graph) {
        const int n = graph.size();
        vector<DirectedGraphNode*> res;
        vector<bool> visited(n, false);
        for (auto pNode : graph) {
            if (!visited[pNode->label])
                dfs(pNode, graph, visited, res);
        }
        reverse(res.begin(), res.end());
        return res;
    }

    void dfs(DirectedGraphNode* ptr, vector<DirectedGraphNode*> &graph,
    vector<bool> &visited, vector<DirectedGraphNode*> &res) {
        visited[ptr->label] = true;
        vector<DirectedGraphNode *> &neigh = ptr->neighbors;
        for (auto pNode : neigh) {
            if (!visited[pNode->label])
                dfs(pNode, graph, visited, res);
        }
        res.push_back(ptr);
    }
};